spiffe/tls_peer.go

```diff
@@ -0,0 +1,319 @@
+	}
+
+	client, err := workload.NewX509SVIDClient(&tlsPeerWatcher{p: p}, dialOpts...)
+	if err != nil {
+		return nil, err
+	}
+	client.Start()
+
+	p.client = client
+	return p, nil
+}
+
+func (p *TLSPeer) Close() error {
+	return p.client.Stop()
+}
+
+func (p *TLSPeer) WaitUntilReady(ctx context.Context) error {
+	select {
+	case <-p.ready:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (p *TLSPeer) GetCertificate() (*tls.Certificate, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	if p.cert == nil {
+		return nil, errors.New("workload does not have a certificate yet")
+	}
+	return p.cert, nil
+}
+
+func (p *TLSPeer) GetRoots() (map[string]*x509.CertPool, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	if p.roots == nil {
+		return nil, errors.New("workload does not have roots yet")
+	}
+	return p.roots, nil
+}
+
+func (p *TLSPeer) updateX509SVIDs(svids *workload.X509SVIDs) {
+	p.log.Debugf("X509SVID workload API update received")
+
+	// Use the default SVID for now
+	// TODO: expand SVID selection options
+	svid := svids.Default()
+	_, trustDomainID, err := internal.GetIDsFromCertificate(svid.Certificates[0])
+	if err != nil {
+		p.onError(errors.New("unable to parse IDs from X509-SVID update"))
+		return
+	}
+
+	cert := &tls.Certificate{
+		Certificate: make([][]byte, 0, len(svid.Certificates)),
+		PrivateKey:  svid.PrivateKey,
+	}
+	for _, svidCert := range svid.Certificates {
+		cert.Certificate = append(cert.Certificate, svidCert.Raw)
+	}
+
+	roots := make(map[string]*x509.CertPool)
+	for federatedDomainID, federatedDomainPool := range svid.FederatedTrustBundlePools {
+		roots[federatedDomainID] = federatedDomainPool
+	}
+	roots[trustDomainID] = svid.TrustBundlePool
+
+	p.mu.Lock()
+	p.cert = cert
+	p.roots = roots
+	p.mu.Unlock()
+
+	p.readyOnce.Do(func() {
+		close(p.ready)
+	})
+}
+
+func (p *TLSPeer) onError(err error) {
+	p.log.Errorf("%v", err)
+}
+
+type tlsPeerWatcher struct {
+	p *TLSPeer
+}
+
+func (w *tlsPeerWatcher) UpdateX509SVIDs(svids *workload.X509SVIDs) {
+	w.p.updateX509SVIDs(svids)
+}
+
+func (w *tlsPeerWatcher) OnError(err error) {
+	w.p.onError(err)
+}
+
+func (p *TLSPeer) Dial(ctx context.Context, network, address string, expectPeer ExpectPeerFunc) (net.Conn, error) {
+	conn, err := new(net.Dialer).DialContext(ctx, network, address)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := p.GetConfig(ctx, expectPeer)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	return tls.Client(conn, config), nil
+}
+
+func (p *TLSPeer) Listen(ctx context.Context, network, address string, expectPeer ExpectPeerFunc) (net.Listener, error) {
+	inner, err := net.Listen(network, address)
+	if err != nil {
+		return nil, err
+	}
+
+	l, err := p.NewListener(ctx, inner, expectPeer)
+	if err != nil {
+		inner.Close()
+		return nil, err
+	}
+
+	return l, nil
+}
+
+func (p *TLSPeer) NewListener(ctx context.Context, inner net.Listener, expectPeer ExpectPeerFunc) (net.Listener, error) {
```

Comment:
What's the distinction between this and Listen? What are the use cases for passing an already-established listener?

TLSPeer.FromListener()?

Reply:
This follows roughly the same call pattern as the function with the same name in the `net/tls`package. It lets you wrap an arbitrary listener (e.g. one implementing a custom transport not supported by the net package).

Reply:
👍